Gems/ROS2/Code/Source/Manipulator/JointPublisherComponent.cpp

```diff
@@ -0,0 +1,118 @@
+{
+    void JointPublisherComponent::Activate()
+    {
+        AZ::TickBus::Handler::BusConnect();
+        auto ros2Node = ROS2::ROS2Interface::Get()->GetNode();
+        AZStd::string jointStatesTopic = "joint_states";
+        m_jointstatePublisher = ros2Node->create_publisher<sensor_msgs::msg::JointState>(jointStatesTopic.data(), rclcpp::QoS(1));        // TODO: add QoS instead of "1"
+    }
+
+    void JointPublisherComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+    }
+
+    void JointPublisherComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided)
+    {
+        provided.push_back(AZ_CRC_CE("JointPublisher"));
+    }
+
+    void JointPublisherComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC_CE("URDFMetadata"));
+    }
+
+    void JointPublisherComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<JointPublisherComponent, AZ::Component>()
+                ->Version(0);
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<JointPublisherComponent>("JointPublisherComponent", "[Publish all the Hinge joint in the tree]")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Game"))
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2");
+            }
+        }
+    }
+
+    void JointPublisherComponent::InitializeMap()
+    {
+        auto* metadataComponent = GetEntity()->FindComponent<URDFMetadataComponent>();
+        m_hierarchyMap = metadataComponent->GetHierarchy();
+        AZ_Printf("JointPublisherComponent", "Map initialized");
+    }
+
+    void JointPublisherComponent::InitializeJointStateMessage()
+    {
+        InitializeMap();
+        double nullPosition{0};
+        for ([[maybe_unused]] auto& [name, entityId] : m_hierarchyMap)
+        {
+            AZ::Entity* hingeEntity = AzToolsFramework::GetEntityById(entityId);
+            AZ_Assert(hingeEntity, "Unknown entity %s", entityId.ToString().c_str());
+            
+            if (hingeEntity->FindComponent<PhysX::HingeJointComponent>() != nullptr)
+            {
+                m_jointstateMsg.name.push_back(name.GetCStr());
+                m_jointstateMsg.position.push_back(nullPosition);
+            }
+        }
+    }
+
+    void JointPublisherComponent::UpdateMessage()
+    {
+        int i = 0;
+        for ([[maybe_unused]] auto& [name, entityId] : m_hierarchyMap)
+        {
+            if (auto* hingeComponent = AzToolsFramework::GetEntityById(entityId)->FindComponent<PhysX::HingeJointComponent>())
+            {
+                m_jointstateMsg.position[i] = GetJointPosition(hingeComponent);
```

Comment:
This logic is unreliable and in principal a security issue as HingeJointComponents may be added/remove at runtime.
A basic fix would be: 
- ensure that `i` does not grow beyond the size of m_hierarchyMap and
- add an assertion after the for loop that warns if `i != size(m_hierarchyMap)`

Reply:
`m_hierarchyMap` is built during the importing of a URDF model, so if you decide to add some joints during runtime it will not be seen by the URDFMetadataComponent (and subsequently by the JointPublisherComponent). Regarding the remotion of a joint from the manipulator chain I do not see any point in doing this, since you are affecting the simulation and the robot model. Beside that, do you think the `i` logic could potentially be risky?

Reply:
If m_hierarchyMap is static at runtime, `i` can indeed not grow beyond the array size. To avoid the potential source of confusion around joints removed at runtime, it should suffice to move the `i++` outside of the `if(...){...}` clause so the remaining joint values will always be written to the correct location. An assertion to warn the users would likely still be a good idea though.